File: 백준/Gold/1107. 리모컨/리모컨.py
import sys

channel = int(sys.stdin.readline())
n = int(sys.stdin.readline())
broken = list(map(int, sys.stdin.readline().split()))

buttons = [True] * 10
for b in broken:
    buttons[b] = False

answer = abs(channel - 100)

def can_make(ch):
    if ch == 0: return buttons[0]
    while ch:
        if not buttons[ch % 10]: return False
        ch //= 10
    return True


def get_close(diff):
    c = channel
    
    while c >= 0 and abs(channel - c) < answer:
        if can_make(c):
            return len(str(c)) + abs(channel - c)
        c += diff
    
    return answer


if len(broken) == 10:
    print(answer)
else:
    print(min(answer, get_close(-1), get_close(1)))

File: 백준/Gold/1107. 리모컨/리모컨.cc
#include <iostream>
#include <string>
#include <vector>

#define FASTIO ios_base::sync_with_stdio(false);cin.tie(NULL);cout.tie(NULL);

using namespace std;

vector<bool> buttons(10, true);

bool can_move(int ch){
    if(ch == 0) return buttons[0];
    while(ch){
        if(!buttons[ch % 10]) return false;
        ch /= 10;
    }
    return true;
}

int main() {
    FASTIO;

    int channel, n;
    cin >> channel;
    cin >> n;
    int broken = 0;
    for(int i = 0; i < n; i++){
        int tmp;
        cin >> tmp;
        buttons[tmp] = false;
        broken++;
    }

    int answer = abs(channel - 100);

    if(broken < 10){
        for(int c = channel; c >= 0 && channel - c < answer; c--){
            if(can_move(c)){
                answer = min(answer, (int)to_string(c).size() + channel - c);
                break;
            }
        }

        for(int c = channel; c - channel < answer; c++){
            if(can_move(c)){
                answer = min(answer, (int)to_string(c).size() + c - channel);
                break;
            }
        }
    }

    cout << answer << endl;

    return 0;
}